pkg/kv/kvserver/replica_raftstorage.go

```diff
@@ -278,27 +262,55 @@ func term(
 	reader storage.Reader,
 	rangeID roachpb.RangeID,
 	eCache *raftentry.Cache,
-	i uint64,
+	index uint64,
 ) (uint64, error) {
-	// entries() accepts a `nil` sideloaded storage and will skip inlining of
-	// sideloaded entries. We only need the term, so this is what we do.
-	ents, err := entries(ctx, rsl, reader, rangeID, eCache, nil /* sideloaded */, i, i+1, math.MaxUint64 /* maxBytes */)
-	if errors.Is(err, raft.ErrCompacted) {
-		ts, err := rsl.LoadRaftTruncatedState(ctx, reader)
-		if err != nil {
-			return 0, err
+	entry, ok := eCache.Get(rangeID, index)
+	if ok {
+		return entry.Term, nil
+	}
+
+	found := false
+	if err := raftlog.Visit(reader, rangeID, index, index+1, func(ent raftpb.Entry) error {
+		if found {
+			return errors.Errorf("found more than one entry in [%d,%d)", index, index+1)
 		}
-		if i == ts.Index {
-			return ts.Term, nil
+		found = true
+		entry = ent
+		return nil
+	}); err != nil {
+		return 0, err
+	}
+
+	if found {
+		// Found an entry. Double-check that it has a correct index.
+		if got, want := entry.Index, index; got != want {
+			return 0, errors.Errorf("there is a gap at index %d, found entry #%d", want, got)
+		}
+		// Cache the entry except if it is sideloaded. We don't load/inline the
+		// sideloaded entries here to keep the term fetching cheap.
+		if !logstore.SniffSideloadedRaftCommand(entry.Data) {
+			eCache.Add(rangeID, []raftpb.Entry{entry}, false /* truncate */)
 		}
-		return 0, raft.ErrCompacted
-	} else if err != nil {
+	}
+	// Otherwise, the entry at the given index is not found. This can happen if
+	// the index is beyond the current log size, or it has been compacted away.
+
+	lastIndex, err := rsl.LoadLastIndex(ctx, reader)
+	if err != nil {
+		return 0, err
+	}
+	if index >= lastIndex {
+		return 0, raft.ErrUnavailable
+	}
+
+	ts, err := rsl.LoadRaftTruncatedState(ctx, reader)
+	if err != nil {
 		return 0, err
 	}
-	if len(ents) == 0 {
-		return 0, nil
+	if index == ts.Index {
```

Comment:
Probably also need to return a "bad" error if index > ts.Index (and < lastIndex). This would mean that there is a gap.

Reply:
Done.